### PR TITLE
[dynamo][api] Better support of torch.nn.Module

### PR DESCRIPTION
**Pull request by anijain2305**
This is an API change, so please review carefully.

With this PR, torchdynamo returns an `OptimizedModule` class object, a subclass of `torch.nn.Module`, when asked to optimize a `nn.Module` object. Most of the methods are redirected to the original `nn.Module`, which is installed as `_mod` in the `OptimizedModule`. 

This is helpful for many cases

```
mod = MockModule()

opt_mod = torch._dynamo.optimize()(mod)

print(opt_mod) # Works

opt_mod = opt_mod.to(device="cuda")
print(opt_mod) # Works
opt_mod(input) # Triggers recompile if necessary, earlier we were shedding the TorchDynamo wrapper

opt_mod.parameters() # Refers to the original module

```

Topics unclear to me
* I have overridden many methods to raise NotImplementedError. A careful review of those will be good.
* hooks
* For the optimized forward, should we call torchdynamo optimization on `__call__` or `forward`
* What else to test

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -188,6 +270,34 @@ def _fn(*args, **kwargs):
         # If the function is called using torch._dynamo.optimize decorator, we
         # should prevent any type of skipping.
         if callback not in (None, False):
+            if not hasattr(fn, "__code__"):
+                raise RuntimeError(
+                    textwrap.dedent(
+                        """
+
+                        torch._dynamo.optimize is called on a non function object.
+                        If this is a callable class, please optimize the individual methods that you are interested in optimizing.
```
Comment (msaroufim) [inline torch/_dynamo/eval_frame.py:279]: Do we want to do this? It feels like a significant step down in UX from where we started i.e: the context manager

Comment (Chillee) [reply torch/_dynamo/eval_frame.py:279]: I'm not sure how this is any worse from what we have now?

**Review by Chillee — COMMENTED** [review]
IMO, we should keep the wrapper simple, and provide an in-place option for users who want a version that less intrusively modifies their module.

File: torch/_dynamo/eval_frame.py (hunk 1 above)
Comment (msaroufim) [reply torch/_dynamo/eval_frame.py:279]: I'm thinking of it as the difference between

`optimize(mod.forward)` and 

`def wrapper(): mod; optimize(wrapper)`

File: torch/_dynamo/eval_frame.py [hunk 2]
```diff
@@ -45,6 +46,103 @@
...
+        return getattr(self._mod, name)
+
+    def forward(self, *args, **kwargs):
+        # This will be monkey patched later
+        raise RuntimeError("Should not be here")
+
+    # Redirect the methods to the original module methods. This avoids things
+    # like `_mod` prefix in parameter names. A more pythonic way might be to use
+    # __getattribute__, but implementing it is very complicated with nn.Module.
+    def parameters(self):
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply torch/_dynamo/eval_frame.py:72]: Do you mean just `parameters()` or other methods as well?

File: torch/_dynamo/eval_frame.py [hunk 3]
```diff
@@ -45,6 +46,103 @@
...
+    def register_buffer(self, name, tensor, persistent=True):
+        self.unimplemented("register_buffer")
+
+    def register_module(self, name, tensor, persistent=True):
+        self.unimplemented("register_module")
+
+    def register_parameter(self, name, tensor, persistent=True):
+        self.unimplemented("register_parameter")
+
+    def state_dict(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply torch/_dynamo/eval_frame.py:141]: Yeah, it sounded like we should support this. I can add the impl.

**Comment by albanD** [comment]
Could you share the justification here on why we want this to be out-of-place and not inplace?

**Comment by mlazos** [comment]
I tried having dynamo trace `__call__` over `forward` as my second task on the team, there were definitely gaps then. As a result I only had dynamo trace `__call__` for lazy modules, which needed the hooks for their initialization. Given, those gaps may be gone now, but I think to be safe `forward` is better.

**Comment by anijain2305** [comment]
> Could you share the justification here on why we want this to be out-of-place and not inplace?

The only reason is that mutation might be unexpected for the user. But, given that the this wrapper impl seems somewhat dirty and can also confuse the user, I am not that much opposed to inplace now.

I would also like to ask @jansel , @ezyang, @voznesenskym @soumith opinion on inplace vs wrapper.

**Comment by ezyang** [comment]
I don't think we should be in the business of module wrapper at all. 

But assuming this is non-negotiable, you will have the least surprise if your module wrapper works the same way that FX/TS/DDP/FSDP module wrapping works, as users have experience with those wrappers.

**Comment by soumith** [comment]
I concur with Ed that I don't think we should be in the business of a module class wrapper.

We should only wrap Module objects.

**Comment by anijain2305** [comment]
> I concur with Ed that I don't think we should be in the business of a module class wrapper.
> 
> We should only wrap Module objects.

What is the difference between module class wrapper and module object wrapper?

More specifically, thats what this PR is doing

~~~
class MockModule(torch.nn.Module):
    .....

mod = MockModule()
opt_mod = torch._dynamo.optimize()(mod)


# opt_mod is of type OptimizedModule. OptimizedModule is a subclass of nn.Module,
# and redirects many of the methods to unoptimized mod methods

# mod is unchanged, so no inplace as of now .We are discussing if we 
# should just do inplace so that all this wrapper redirection can be avoided.
~~~

**Review by jansel — COMMENTED** [review]
+1 on handling module instances rather than classes.  I believe that is how it is done in this PR already, so no change needed.

Where there is a choice, I'd also lean towards consistency with FX/TS.

From a design perspective, I think there is a key choice to make:
1) Is the wrapper module's goal to be entirely transparent to the user and masquerade as the original module
2) Or, is the wrapper module user-visible and more straightforward

It seems like the implementation is trying to be closer to option 1, where things like `str` and `parameters` hide the fact there is a wrapper.  Though the wrapper-ness will leak out in lots of places (e.g. unimplemented methods, `isinstance`, etc) -- so it is an imperfect masquerade job.

I am torn on which way to go.  I like option 2 for its simplicity, we barely need to override anything, though it might require more changes to user code if people pass modules as args to things.

If we want to do option 1, I think we could do better and fake it in a more convincing way.  We could shallow copy the module then monkey patch it, etc.  This could lead to some BC issues later on though, since users will start relying on this behavior.

I'd like be grounded in user needs here.  Is option 1 something users are asking for?  What are the use cases they have?  How common is it?

**Comment by anijain2305** [comment]
@jansel Thanks for de-cluttering the discussion.

I think we want following cases to work well

Case 1 
```
mod = MockModule()
opt_mod = torch._dynamo.optimize()(mod)
opt_mod = opt_mod.cuda()
opt_mod(x)
```


Case 2 
```
mod = MockModule()
opt_mod = torch._dynamo.optimize()(mod)

opt_mod(x)
```

**Comment by anijain2305** [comment]
@jansel Thanks for providing the clarity.

I think we want following 2 cases to behave same from user perspective

Case 1 
```
mod = MockModule()
mod = mod.cuda()
opt_mod = torch._dynamo.optimize()(mod)
opt_mod(x)
print(type(opt_mod))
>> <class 'torch._dynamo.eval_frame._TorchDynamoContext.__call__.<locals>.TorchDynamoNNModuleWrapper'>

```


Case 2 
```
mod = MockModule()
opt_mod = torch._dynamo.optimize()(mod)
opt_mod = opt_mod.cuda()
opt_mod(x)
print(type(opt_mod))
>> <class '__main__.MockModule'>
```
But on master branch, case 2 will shed the TorchDynamo wrapper given the simplicity of our wrapper.

So, I am going for design option 1 that you mentioned in your message
   1. Is the wrapper module's goal to be entirely transparent to the user and masquerade as the original module

However, as it's difficult to do a perfect masquerading job, maybe we don't have to masquerade. We could have a very simple `nn.Module` wrapper, which just installs the original module (just like this PR does), but does not redirect to the original module. In that case, user will have to be ok to see `_mod` (better names suggestions welcomed) in the parameter names.

cc @Chillee

**Comment by msaroufim** [comment]
Can I ask a dumb question, why do we have an API that works over instances of nn.modules to start with? 

`opt_module = torchdynamo.optimize("inductor")(module)` is strongly hinting that the optimized module is a `nn.Module` not a wrapper.

Whereas if users can only pass in functions this ambiguity goes away, of course I can print stuff, save stuff or do whatever I want it's an `nn.Module`

```python
@torchdynamo.optimize("inductor")
def user_code():
  module = module.cuda()
  print(module)
  torch.save(module)
```

**Comment by anijain2305** [comment]
> Can I ask a dumb question, why do we have an API that works over instances of nn.modules to start with?

It is a very valid question. Imo, we want this because PyTorch users are familiar with modules. Suppose, you want to do inference, this is what the code will look like w/ and w/o explicitly supporting `nn.Module` instances

w/ nn.Module support
~~~
mod = MockModule()
opt_mod = torch._dynamo.optimize()(mod)
mod(x)
~~~

w/o nn.Module support
~~~
mod = MockModule()

# Users have to write this wrapper
def wrapper(*args, **kwargs):
     return mod(*args, **kwargs)

opt_wrapper = torch._dynamo.optimize()(wrapper)
opt_wrapper(x)
~~~

We are trying to avoid that inconvenience of writing a wrapper.

**Comment by msaroufim** [comment]
Thanks @anijain2305 I think that wrapper looks gross if you assume that user code will just have a single function `mod.forward()`

But a more realistic scenario for inference is something like

```python
def inference(model, binary_data)
  data = preprocess(binary_data)
  pred = model(data)
  output = postprocess(pred)
```

And for training you'd have something like

```python
def train(model,data):
  for batch in data:
    model(data)
...

In which case the wrapper functions already exist its just `inference()` or `train()` so my point is that I'd expect for most user code that a wrapper would be much less of an inconvenience relative to learning about a new wrapper.

**Comment by Chillee** [comment]
@jansel 

> I am torn on which way to go. I like option 2 for its simplicity, we barely need to override anything, though it might require more changes to user code if people pass modules as args to things.

I strongly vote for option 2, and do an even simpler module wrapper than what we have in the PR. Basically, we should not pretend to be transparent, and make it explicit that we're doing a module wrapper (although we should still make things convenient, hence why an `nn.module` wrapper). This option requires low amounts of magic, and should be fairly understandable for users (although it may affect their code).

I'd note that it is *very easy for users to do option 1 themselves*. They can just do `mod.forward = torch._dynamo.optimize()(mod.forward)`, and in the future, we can plausibly add a `mod.compile()`. So, considering that it's always easier to go from "more restrictive" => "less restrictive", I vote for option 2.

@msaroufim 

> I fee like a function wrapper will be a lot less annoying to work with than a module wrapper

To be clear, we *do* still support a function wrapper. But many users will also expect to be able to optimize a module, hence this discussion.

**Comment by anijain2305** [comment]
@Chillee Yes, I like your proposal. This prevents us from mutating the module. I updated the PR to reflect the changes.

**Review by Chillee — APPROVED** [review]
Looks good to me overall, although not super certain on the method of checking whether an object is a function/naming concerns.

**Comment by anijain2305** [comment]
@jansel @ezyang @msaroufim This is ready for review (1 failing test is broken on trunk)

**Review by jansel — APPROVED** [review]

**Review by msaroufim — APPROVED** [review]

**Comment by msaroufim** [comment]
Stamping since this is better than we have today but we should monitor tickets people open and see if it becomes a source of confusion and if so we should reevaluate whether we need to be optimizing modules directly. I may just be overly paranoid and maybe noone will even notice lol

**Comment by ezyang** [comment]
So, there's at least one problem with this PR, which is that deepcopy is not implemented correctly. Without this patch, this script passes, but with it, it fails:

```
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch._dynamo.debug_utils import same_two_models

args = [((2, 16, 196, 128), (401408, 25088, 128, 1), torch.float32, 'cuda', True), ((2, 16, 196, 512), (1605632, 100352, 512, 1), torch.float32, 'cuda', True)]
args = [rand_strided(sh, st, dt, dev).requires_grad_(rg) for (sh, st, dt, dev, rg) in args]


from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.self_self_levels_0_transformer_encoder_1_mlp_fc2 = Linear(in_features=512, out_features=128, bias=True)

    def forward(self, add_3, self_self_levels_0_transformer_encoder_1_mlp_drop1):
        self_self_levels_0_transformer_encoder_1_mlp_fc2 = self.self_self_levels_0_transformer_encoder_1_mlp_fc2(self_self_levels_0_transformer_encoder_1_mlp_drop1);  self_self_levels_0_transformer_encoder_1_mlp_drop1 = None
        return self_self_levels_0_transformer_encoder_1_mlp_fc2

mod = Repro().cuda()
opt_mod = torch._dynamo.optimize("eager")(mod)


mod.eval()
opt_mod.eval()
with torch.cuda.amp.autocast(enabled=False):
    assert same_two_models(mod, mod, args), "Eager itself failed"
    assert same_two_models(mod, opt_mod, args), "Dynamo failed"
```

The problem is that when we deepcopy the optimized module, the parameters that we end up checking in same_two_models are different from what parameters are actually used inside the module. There must be something going on where we are swapping out the parameters on OptimizedModule, but not on the inner orig module, where the parameters are actually accessed.

h/t @voznesenskym for finding the minimal repro

**Comment by ezyang** [comment]
btw, this is another reason I hate dynamo plopping full nn modules into the traced fx graph with the fury of a thousand suns. So much complication for so little gain (cc @wconstab)

**Comment by ezyang** [comment]
also, fyi @anjali411 forward fixed the test failure by skipping it on a recent PR she landed, make sure you remove the skip when validating a fix. I'll also add the minimal repro to the test suite

**Pull request closed without merging**